crates/xilem_html/src/element/attributes.rs

```diff
@@ -0,0 +1,126 @@
+use std::borrow::Cow;
+
+use crate::vecmap::VecMap;
+
+type CowStr = Cow<'static, str>;
+
+#[derive(PartialEq, Debug)]
+pub enum AttributeValue {
+    Null,
+    True, // for the boolean true, this serializes to an empty string (e.g. for <input checked>)
+    I32(i32),
+    U32(u32),
+    F32(f32),
+    F64(f64),
+    String(CowStr),
+}
+
+impl AttributeValue {
+    pub fn serialize(&self) -> CowStr {
+        match self {
+            AttributeValue::True => "".into(), // empty string is equivalent to a true set attribute
+            AttributeValue::I32(n) => n.to_string().into(),
+            AttributeValue::U32(n) => n.to_string().into(),
+            AttributeValue::F32(n) => n.to_string().into(),
+            AttributeValue::F64(n) => n.to_string().into(),
+            AttributeValue::String(s) => s.clone(),
+            // AttributeValue::Null shouldn't be serialized within attribute serialization/diffing
+            // as null values should never show up in the attributes map, but for completeness it's serialized to an empty string
+            AttributeValue::Null => "".into(),
+        }
+    }
+}
+
+// A few convenient From implementations for less boilerplate when setting attributes
+
+impl<T: Into<AttributeValue>> From<Option<T>> for AttributeValue {
+    fn from(value: Option<T>) -> Self {
+        if let Some(value) = value {
+            value.into()
+        } else {
+            AttributeValue::Null
+        }
+    }
+}
+
+impl From<bool> for AttributeValue {
+    fn from(value: bool) -> Self {
+        if value {
+            AttributeValue::True
+        } else {
+            AttributeValue::Null
+        }
+    }
+}
+
+impl From<u32> for AttributeValue {
+    fn from(value: u32) -> Self {
+        AttributeValue::U32(value)
+    }
+}
+
+impl From<i32> for AttributeValue {
+    fn from(value: i32) -> Self {
+        AttributeValue::I32(value)
+    }
+}
+
+impl From<f32> for AttributeValue {
+    fn from(value: f32) -> Self {
+        AttributeValue::F32(value)
+    }
+}
+
+impl From<f64> for AttributeValue {
+    fn from(value: f64) -> Self {
+        AttributeValue::F64(value)
+    }
+}
+
+impl From<String> for AttributeValue {
+    fn from(value: String) -> Self {
+        AttributeValue::String(value.into())
+    }
+}
+
+impl From<CowStr> for AttributeValue {
+    fn from(value: CowStr) -> Self {
+        AttributeValue::String(value)
+    }
+}
+
+impl From<&'static str> for AttributeValue {
+    fn from(value: &'static str) -> Self {
+        AttributeValue::String(value.into())
+    }
+}
+
+#[derive(Default)]
+pub struct Attributes(VecMap<CowStr, AttributeValue>);
```

Comment:
If `Null` is never held here, does it really make sense to be part of the enum? I think it makes sense for the invariant to be enforced properly, though this probably means having to use a custom `ToAttributeValue` trait if we want `insert` to be able to unset an attribute value they we it can now.